File: ios/HealthTracker/Services/FunnelAnalytics.swift
import Foundation

/// Minimal, privacy-respecting funnel analytics.
///
/// Fire-and-forget anonymous events to the first-party Vercel endpoint so we can
/// see WHERE users drop in the paywall / purchase funnel (the app is otherwise
/// account-free and unmonetized-by-tracking). No PII is sent: only the anonymous
/// per-install UUID already used for meal-scan rate limiting, the event name, an
/// optional context label, the platform, and a timestamp. It never blocks the UI
/// and silently ignores failures.
final class FunnelAnalytics {
    static let shared = FunnelAnalytics()
    private init() {}

    enum Event: String {
        case onboardingComplete = "onboarding_complete"
        case firstScan          = "first_scan"
        case paywallShown       = "paywall_shown"
        case paywallDismissed   = "paywall_dismissed"
        case buyTapped          = "buy_tapped"
        case purchaseSuccess    = "purchase_success"
        case purchaseFailed     = "purchase_failed"
        case screenView         = "screen_view"
    }

    /// Screens reported through `screenView`. The raw value is the `context`
    /// sent with the event. Mirrored on Android by `FunnelAnalytics.Screen`;
    /// keep the shared names identical on both platforms. Onboarding steps are
    /// per platform because the two onboarding flows differ.
    enum Screen: String {
        case dashboard            = "dashboard"
        case diary                = "diary"
        case addToDiary           = "add_to_diary"
        case mealScanner          = "meal_scanner"
        case mealPlan             = "meal_plan"
        case more                 = "more"
        case paywall              = "paywall"
        case onboardingWelcome    = "onboarding_welcome"
        case onboardingQuickSetup = "onboarding_quick_setup"
        case onboardingReminders  = "onboarding_reminders"
    }

    /// Same domain as the meal-scan proxy; events land at /api/v1/event.
    private let endpoint = URL(string: "https://mochasmindlab.com/api/v1/event")

    private var screenThrottle = ScreenViewThrottle()

    func log(_ event: Event, trigger: PaywallTrigger?) {
        log(event, context: trigger?.analyticsName)
    }

    func log(_ event: Event, context: String? = nil) {
        guard let url = endpoint else { return }

        var body: [String: Any] = [
            "event": event.rawValue,
            "ts": Int(Date().timeIntervalSince1970),
        ]
        if let context = context { body["context"] = context }
        guard let data = try? JSONSerialization.data(withJSONObject: body) else { return }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(SecretsManager.appSharedSecret, forHTTPHeaderField: "X-App-Secret")
        req.setValue(SecretsManager.installId, forHTTPHeaderField: "X-Install-Id")
        req.setValue("ios", forHTTPHeaderField: "X-Platform")
        req.httpBody = data
        req.timeoutInterval = 8

        // Fire-and-forget: analytics must never block, retry, or crash the app.
        URLSession.shared.dataTask(with: req).resume()
    }

    /// Sends `event` at most once per install.
    ///
    /// The top-of-funnel milestones are "first time" events: `onboardingComplete`
    /// marks an activated install and `firstScan` marks the first meal actually
    /// scanned, so both are denominators: one per install or they count nothing.
    /// Onboarding has three exits and the scanner runs many times, so the guard
    /// lives here rather than at each call site. The flag is a plain UserDefaults
    /// bool, so it is wiped by an uninstall or a Clear Data, exactly like the
    /// anonymous install ID it is reported against. Mirrored on Android by
    /// `FunnelAnalytics.logOnce`.
    func logOnce(_ event: Event, context: String? = nil) {
        let key = "funnel_logged_\(event.rawValue)"
        guard !UserDefaults.standard.bool(forKey: key) else { return }
        UserDefaults.standard.set(true, forKey: key)
        log(event, context: context)
    }

    /// Records one `screenView` per appearance of `screen`.
    ///
    /// Call it from the screen's `onAppear`. SwiftUI can fire `onAppear` more
    /// than once for a single visible appearance (tab restores, sheet dismissals
    /// re-presenting the view underneath), so repeats of the same screen inside
    /// a two second window are dropped. A different screen always logs.
    /// Mirrored on Android by `FunnelAnalytics.logScreen`.
    func logScreen(_ screen: Screen) {
        guard screenThrottle.admit(screen.rawValue) else { return }
        log(.screenView, context: screen.rawValue)
    }
}

/// Drops repeats of the same screen name that arrive within `window` seconds of
/// the last one that was admitted. Pure value type so the rule is unit-testable
/// without touching the network.
struct ScreenViewThrottle {
    static let defaultWindow: TimeInterval = 2

    let window: TimeInterval
    private var lastScreen: String?
    private var lastAdmittedAt: Date?

    init(window: TimeInterval = ScreenViewThrottle.defaultWindow) {
        self.window = window
    }

    /// Returns true when `screen` should be logged, and remembers it if so.
    mutating func admit(_ screen: String, now: Date = Date()) -> Bool {
        if let last = lastScreen, let at = lastAdmittedAt,
           last == screen, now.timeIntervalSince(at) < window {
            return false
        }
        lastScreen = screen
        lastAdmittedAt = now
        return true
    }
}

extension PaywallTrigger {
    /// Stable snake_case label for analytics (decoupled from UI copy).
    var analyticsName: String {
        switch self {
        case .mealScanner:         return "meal_scanner"
        case .mealScannerLastScan: return "meal_scanner_last_scan"
        case .mealPlan:            return "meal_plan"
        case .recipeBook:          return "recipe_book"
        case .barcodeScanner:      return "barcode_scanner"
        case .supplements:         return "supplements"
        case .fasting:             return "fasting"
        case .settings:            return "settings"
        case .general:             return "general"
        }
    }
}

File: android/app/src/main/java/com/mochasmindlab/mlhealth/services/FunnelAnalytics.kt
package com.mochasmindlab.mlhealth.services

import com.mochasmindlab.mlhealth.utils.PreferencesManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Minimal, privacy-respecting funnel analytics (mirrors iOS FunnelAnalytics).
 *
 * Fire-and-forget anonymous events so we can see WHERE users drop in the paywall
 * / purchase funnel. No PII: reuses the meal-scan APP_SHARED_SECRET + the
 * anonymous per-install UUID, plus event name, optional context, platform, ts.
 * Never blocks the UI and silently ignores failures. Events land at
 * mochasmindlab.com/api/v1/event.
 */
@Singleton
class FunnelAnalytics @Inject constructor(
    private val prefs: PreferencesManager,
) {
    enum class Event(val value: String) {
        ONBOARDING_COMPLETE("onboarding_complete"),
        FIRST_SCAN("first_scan"),
        PAYWALL_SHOWN("paywall_shown"),
        PAYWALL_DISMISSED("paywall_dismissed"),
        BUY_TAPPED("buy_tapped"),
        PURCHASE_SUCCESS("purchase_success"),
        PURCHASE_FAILED("purchase_failed"),
        SCREEN_VIEW("screen_view"),
    }

    /**
     * Screens reported through [logScreen]. [value] is the `context` sent with
     * the event. Mirrors iOS `FunnelAnalytics.Screen`; keep the shared names
     * identical on both platforms. Onboarding steps are per platform because
     * the two onboarding flows differ (iOS: welcome, quick setup, reminders).
     */
    enum class Screen(val value: String) {
        DASHBOARD("dashboard"),
        DIARY("diary"),
        ADD_TO_DIARY("add_to_diary"),
        MEAL_SCANNER("meal_scanner"),
        MEAL_PLAN("meal_plan"),
        MORE("more"),
        PAYWALL("paywall"),
        ONBOARDING_WELCOME("onboarding_welcome"),
        ONBOARDING_BASIC_INFO("onboarding_basic_info"),
        ONBOARDING_BODY_METRICS("onboarding_body_metrics"),
        ONBOARDING_ACTIVITY_LEVEL("onboarding_activity_level"),
        ONBOARDING_GOALS("onboarding_goals"),
    }

    private val client = OkHttpClient()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val jsonMedia = "application/json".toMediaType()
    private val screenThrottle = ScreenViewThrottle()

    fun log(event: Event, context: String? = null) {
        scope.launch {
            try {
                val secret = SecretsManager.appSharedSecret ?: return@launch
                val installId = prefs.getOrCreateInstallId()
                val payload = JSONObject().apply {
                    put("event", event.value)
                    put("ts", System.currentTimeMillis() / 1000)
                    if (context != null) put("context", context)
                }.toString()
                val req = Request.Builder()
                    .url("https://mochasmindlab.com/api/v1/event")
                    .addHeader("X-App-Secret", secret)
                    .addHeader("X-Install-Id", installId)
                    .addHeader("X-Platform", "android")
                    .post(payload.toRequestBody(jsonMedia))
                    .build()
                client.newCall(req).execute().use { /* fire-and-forget */ }
            } catch (_: Exception) {
                // Analytics must never block, retry, or crash the app.
            }
        }
    }

    /**
     * Sends [event] at most once per install.
     *
     * The top-of-funnel milestones are "first time" events: ONBOARDING_COMPLETE
     * marks an activated install and FIRST_SCAN marks the first meal actually
     * scanned, so both are denominators: one per install or they count nothing.
     * The scanner runs many times, so the guard lives here rather than at the
     * call site. The flag is a DataStore bool, wiped by an uninstall or Settings
     * -> Clear Data, exactly like the anonymous install ID it is reported
     * against. Mirrors iOS `FunnelAnalytics.logOnce`.
     */
    fun logOnce(event: Event, context: String? = null) {
        scope.launch {
            try {
                if (prefs.markFunnelEventLogged(event.value)) {
                    log(event, context)
                }
            } catch (_: Exception) {
                // Same contract as log(): never block, retry, or crash.
            }
        }
    }

    /**
     * Records one SCREEN_VIEW per appearance of [screen].
     *
     * Compose can re-enter a screen for a single visible appearance (a
     * configuration change recreates the composition, a bottom sheet closing
     * over a tab), so repeats of the same screen inside a two second window are
     * dropped. A different screen always logs. Mirrors iOS
     * `FunnelAnalytics.logScreen`.
     */
    fun logScreen(screen: Screen) {
        if (!screenThrottle.admit(screen.value)) return
        log(Event.SCREEN_VIEW, screen.value)
    }
}
